File: PoketDex/app/src/main/java/pt/rfsfernandes/ui/fragments/details_pager_fragments/MovesInfoFragment.java
package pt.rfsfernandes.ui.fragments.details_pager_fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import pt.rfsfernandes.custom.Constants;
import pt.rfsfernandes.custom.adapters.PokemonMovesAdapter;
import pt.rfsfernandes.custom.callbacks.MovesItemClick;
import pt.rfsfernandes.custom.dialogs.SimpleCustomDialog;
import pt.rfsfernandes.databinding.FragmentMovesInfoBinding;
import pt.rfsfernandes.ui.activities.MainActivity;
import pt.rfsfernandes.viewmodels.MainViewModel;

public class MovesInfoFragment extends Fragment implements MovesItemClick {

  private FragmentMovesInfoBinding binding;
  private MainViewModel mMainViewModel;
  private PokemonMovesAdapter mPokemonMovesAdapter;

  public MovesInfoFragment() {
    // Required empty public constructor
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
                           Bundle savedInstanceState) {
    binding = FragmentMovesInfoBinding.inflate(inflater, container, false);
    View view = binding.getRoot();
    mMainViewModel = new ViewModelProvider(requireActivity()).get(MainViewModel.class);
    mPokemonMovesAdapter = new PokemonMovesAdapter(getContext(), this);

    LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext());
    linearLayoutManager.setOrientation(RecyclerView.HORIZONTAL);
    binding.recyclerViewMovesInfo.setLayoutManager(linearLayoutManager);
    binding.recyclerViewMovesInfo.setAdapter(mPokemonMovesAdapter);
    return view;
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    initViewModel();
  }

  /**
   * Initiates the viewModel observers
   */
  private void initViewModel() {
    mMainViewModel.getPokemonMutableLiveData().observe(getViewLifecycleOwner(), pokemon -> {
      mMainViewModel.isLoading(true);
      mMainViewModel.getMovesFromIds(pokemon.getMoveslist());
    });

    mMainViewModel.getMovesInfo().observe(getViewLifecycleOwner(), moves -> {
      mPokemonMovesAdapter.refreshList(moves);
      mMainViewModel.isLoading(false);
    });
  }

  @Override
  public void clickedMovesItem(String title, String text, Constants.MOVES_ITEM moves_item) {
    mMainViewModel.isLoading(true);
    switch (moves_item) {
      case TYPE:
        if (getActivity() != null) {
          mMainViewModel.isLoading(true);
          ((MainActivity) getActivity()).setShowTypeInfo(true);
          mMainViewModel.getTypeAndCounters(Integer.parseInt(text), Constants.SHOW_TYPE.MOVE);
        }
        break;
      case SIMPLE:
        showSimpleDialogContent(title, text);
        break;
    }
  }

  /**
   * Shows a simple custom dialog with a title and a content
   *
   * @param title   Title of the dialog
   * @param content Content of the dialog
   */
  private void showSimpleDialogContent(String title, String content) {
    SimpleCustomDialog simpleCustomDialog = new SimpleCustomDialog(getActivity());
    simpleCustomDialog.show();
    simpleCustomDialog.getTextViewSimpleTitle().setText(title);
    simpleCustomDialog.getTextViewSimpleText().setText(content);
    mMainViewModel.isLoading(false);
  }

}

File: PoketDex/app/src/main/java/pt/rfsfernandes/pocketdex/ui/fragments/details_pager_fragments/MovesInfoFragment.kt
package pt.rfsfernandes.pocketdex.ui.fragments.details_pager_fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import pt.rfsfernandes.pocketdex.custom.Constants
import pt.rfsfernandes.pocketdex.custom.Constants.MOVES_ITEM
import pt.rfsfernandes.pocketdex.custom.adapters.PokemonMovesAdapter
import pt.rfsfernandes.pocketdex.custom.callbacks.MovesItemClick
import pt.rfsfernandes.pocketdex.custom.dialogs.SimpleCustomDialog
import pt.rfsfernandes.pocketdex.databinding.FragmentMovesInfoBinding
import pt.rfsfernandes.pocketdex.model.moves.Moves
import pt.rfsfernandes.pocketdex.ui.activities.MainActivity
import pt.rfsfernandes.pocketdex.viewmodels.MainViewModel

class MovesInfoFragment : Fragment(), MovesItemClick {
    private lateinit var binding: FragmentMovesInfoBinding
    private val mMainViewModel: MainViewModel by sharedViewModel()
    private lateinit var mPokemonMovesAdapter: PokemonMovesAdapter
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentMovesInfoBinding.inflate(inflater, container, false)
        val view: View = binding.root
        mPokemonMovesAdapter = PokemonMovesAdapter(requireContext(), this)
        val linearLayoutManager = LinearLayoutManager(context)
        linearLayoutManager.orientation = RecyclerView.HORIZONTAL
        binding.recyclerViewMovesInfo.layoutManager = linearLayoutManager
        binding.recyclerViewMovesInfo.adapter = mPokemonMovesAdapter
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViewModel()
    }

    /**
     * Initiates the viewModel observers
     */
    private fun initViewModel() {
        mMainViewModel.pokemonMutableLiveData.observe(viewLifecycleOwner) { (_, _, _, moveslist) ->
            mMainViewModel.isLoading(true)
            mMainViewModel.getMovesFromIds(moveslist)
        }
        mMainViewModel.movesInfo.observe(viewLifecycleOwner) { moves: List<Moves?>? ->
            moves?.let { mPokemonMovesAdapter.refreshList(it) }
            mMainViewModel.isLoading(false)
        }
    }

    /**
     * Shows a simple custom dialog with a title and a content
     *
     * @param title   Title of the dialog
     * @param content Content of the dialog
     */
    private fun showSimpleDialogContent(title: String, content: String) {
        val simpleCustomDialog = SimpleCustomDialog(requireActivity())
        simpleCustomDialog.show()
        simpleCustomDialog.textViewSimpleTitle?.text = title
        simpleCustomDialog.textViewSimpleText?.text = content
        mMainViewModel.isLoading(false)
    }

    override fun clickedMovesItem(title: String?, text: String?, moves_item: MOVES_ITEM?) {
        mMainViewModel.isLoading(true)
        when (moves_item) {
            MOVES_ITEM.TYPE -> if (activity != null) {
                mMainViewModel.isLoading(true)
                (activity as MainActivity?)!!.setShowTypeInfo(true)
                text?.toInt()
                    ?.let { mMainViewModel.getTypeAndCounters(it, Constants.SHOW_TYPE.MOVE) }
            }
            MOVES_ITEM.SIMPLE -> {
                if(title != null && text != null)
                    showSimpleDialogContent(title, text)
            }
        }
    }
}